clippy_lints/src/unused_io_amount.rs

```diff
@@ -82,37 +85,74 @@ impl<'tcx> LateLintPass<'tcx> for UnusedIoAmount {
         }
 
         if let Some(exp) = block.expr
-            && matches!(exp.kind, hir::ExprKind::If(_, _, _) | hir::ExprKind::Match(_, _, _))
+            && matches!(
+                exp.kind,
+                hir::ExprKind::If(_, _, _) | hir::ExprKind::Match(_, _, hir::MatchSource::Normal)
+            )
         {
             check_expr(cx, exp);
         }
     }
 }
 
+fn non_consuming_err_arm<'a>(cx: &LateContext<'a>, arm: &hir::Arm<'a>) -> ControlFlow<&'a hir::Pat<'a>, ()> {
+    // if there is a guard, we consider the result to be consumed
+    if arm.guard.is_some() {
+        return ControlFlow::Continue(());
+    }
+    if is_unreachable_or_panic(cx, arm.body) {
+        // if the body is unreachable or there is a panic,
+        // we consider the result to be consumed
+        return ControlFlow::Continue(());
+    }
+
+    if let PatKind::TupleStruct(ref path, [inner_pat], _) = arm.pat.kind
+        // we check against Result::Ok to avoid linting on Err(_) or something else.
+        && is_res_lang_ctor(cx, cx.qpath_res(path, inner_pat.hir_id), hir::LangItem::ResultErr)
+    {
+        return ControlFlow::Break(inner_pat);
+    }
+
+    ControlFlow::Continue(())
+}
+
+fn non_consuming_ok_arm<'a>(cx: &LateContext<'a>, arm: &hir::Arm<'a>) -> ControlFlow<&'a hir::Pat<'a>, ()> {
+    // if there is a guard, we consider the result to be consumed
+    if arm.guard.is_some() {
+        return ControlFlow::Continue(());
+    }
+    if is_unreachable_or_panic(cx, arm.body) {
+        // if the body is unreachable or there is a panic,
+        // we consider the result to be consumed
+        return ControlFlow::Continue(());
+    }
+
+    if let Some(OkPatternKind::WildOrDotDot) = extract_ok_pattern(cx, arm.pat) {
+        return ControlFlow::Break(arm.pat);
+    }
```

Comment:
With that previous comment, this could be:

```suggestion
    return extract_ok_pattern(cx, arm.pat);
```

This would also need to return a bool. But as this is the only the branch that returns meaningful information, making the switch won't be hard I'd suppose.

Reply:
If we return a bool here, then we lose information on which arm is doing what.

These are used to pull the right patterns out of a slice of 2 pats.